### ISSUE TITLE
A disposed Provider does not remove itself from the dependents of another provider during frequent updates

### ISSUE THREAD

**Issue by Quijx** [rrousselGit/riverpod#1385]
## Describe the bug
When an autodisposed Provider is disposed on closing a page it is used in, while its value changes frequently causes the provider it depends on to keep the disposed provider as a dependent. This causes it to recompute its value even though it is not a dependency of any other provider or widget.
(More specific details in the Analysis section)

Riverpod version: `2.0.0-dev.5`
Flutter version: `2.10.5`

The app in the following minimal reproducible example shows an app that fakes loading data on the press of a "Go" button. However the loaded data depends on some other data, in this case the value of a counter, which can be be increased with a "+" FAB. The desired behavior is that the data is loaded only once it is viewed, and not when the counter changes while it is not being viewed.

![riverpod_bug](https://user-images.githubusercontent.com/48153628/165591943-2ccee574-ceeb-4ef1-b305-9aefdcedfd63.gif)

## To Reproduce
1. Run `flutter create riverpod_bug`, add `flutter_riverpod: ^2.0.0-dev.5` to the `dependencies` in the `pubspec.yaml` and copy the following code from the minimal reproducible example into the `main.dart`
<details>
  <summary>Expand minimal reproducible example</summary>
  
```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

const totalProgress = 100;

final counterProvider = StateProvider<int>((ref) => 0);

final progressProvider = StreamProvider<int>((ref) async* {
  ref.onCancel(() => print('progressProvider cancel'));
  ref.watch(counterProvider);
  for (var i = 0; i <= totalProgress; i++) {
    // Sums to a total of about 2 seconds.
    await Future.delayed(const Duration(milliseconds: 2000 ~/ totalProgress));
    yield i;
  }
});

final progressCopyProvider = Provider.autoDispose<AsyncValue<int>>((ref) {
  ref.onDispose(() => print('progressCopyProvider dispose'));
  return ref.watch(progressProvider);
});

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return const ProviderScope(
      child: MaterialApp(title: 'Riverpod Bug', home: HomePage()),
    );
  }
}

class HomePage extends ConsumerWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      appBar: AppBar(title: const Text('Riverpod Bug')),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text('You have pushed the button this many times:'),
            Text(
              ref.watch(counterProvider).toString(),
              style: Theme.of(context).textTheme.headline4,
            ),
            ElevatedButton(
              onPressed: () {
                Navigator.of(context).push(
                  MaterialPageRoute(
                    builder: (context) => const LoadingPage(),
                  ),
                );
              },
              child: const Text('Go'),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          ref.read(counterProvider.notifier).state++;
        },
        child: const Icon(Icons.add),
      ),
    );
  }
}

class LoadingPage extends ConsumerWidget {
  const LoadingPage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return Scaffold(
      appBar: AppBar(title: const Text('Loading')),
      body: Center(
        child: ref.watch(progressCopyProvider).maybeWhen(
              data: (i) {
                if (i == totalProgress) {
                  return const Text('Done');
                }
                return CircularProgressIndicator(
                  value: i.toDouble() / totalProgress,
                );
              },
              orElse: () => const Text('Else'),
            ),
      ),
    );
  }
}
```
</details>

2. Tap the "Go" button. A loading screen with a filling progress indicator appears.
3. Wait until the loading completes and tap the back button in the top left to go back to the home screen.
4. Tap the "+" FAB on the home screen and wait a few seconds.
5. Tap the "Go" button again. The loading screen appears again and loads starting from 0.
  - This is as expected because the `progressProvider` watches the `counterProvider`. And since the count changed, the loading needs to restart from scratch.
  - However most notably, the loading only starts once the loading screen is opened, because the `progressCopyProvider` is `autoDispose` and was disposed after the loading screen was closed previously. Therefore the `progressProvider` didn't bother reloading because no one was watching until now.
6. Tap the back button in the top left, before the loading completes.
7. Tap the "+" FAB on the home screen
8. One second later tap the "Go" button again.

### Expected behavior

- The loading screen should again start from 0 because the `progressCopyProvider` marked as `autoDispose` should have removed itself as dependant of the `progressProvider` after closing the loading screen during the loading. And since again no one is watching `progressProvider`, it should not have started the loading again.

### Actual behavior

- The loading screen opens while the progress was counting up already. (Or if you wait 3 seconds after pushing the "Go", the loading screen starts at "Done" already.)
- This is because the `progressProvider` still thinks the `progressCopyProvider` is listening (even though it's disposed). And reran the create callback the moment the "+" FAB was pressed, because it has it dependency on the `counterProvider`.

## Analysis

My search for the bug so far has shown that this happens because during the frequent updates in the loading screen, the `AutoDisposeProviderElement` of the `progressCopyProvider` is mounted and dismounted repeatedly. And when the back button is pressed, the `_ProviderScheduler` basically always calls `_performDispose` during a time at which the `AutoDisposeProviderElement` for the `progressCopyProvider` is not mounted. This causes the check at
https://github.com/rrousselGit/riverpod/blob/b1a7128fabe6862632899c8764ad396ad42f44e1/packages/riverpod/lib/src/framework/scheduler.dart#L79-L84
to continue and skip the provider because the `!element._mounted` is `true`.

The two prints in the providers also give some insight into this behavior. If the loading page is closed while done, first `progressCopyProvider dispose` and then `progressProvider cancel` is printed. If it is closed while loading, only `progressCopyProvider dispose` is printed a lot of times (during the updates).

If the `const totalProgress = 100;` at the top is set to for example `10`, the problem does not happen, because the provider is mounted most of the time. (Maybe it sometimes happens if you get unlucky and press "back" just as an update comes in?)

I don't know what the actual root cause of this problem is. Should the provider even be dismounted and mounted every time the value changes? Should the scheduler only call `_performDispose` at a time where the provider is mounted? Or should the check omit the test for if the provider element is mounted?

**Comment by rrousselGit** [comment]
Thanks a lot for the very detailed report! I'll investigate

**Comment by rrousselGit** [comment]
It sounds like the _mounted check in the scheduler is completely unnecessary and is the one causing the problem.

Thinking about it, that _mounted check doesn't make sense as even if the provider is currently disposed of, we still want to remove its element

**Comment by rrousselGit** [comment]
So I found a reliable way to reproduce the problem and fixed it.

This has revealed a separate issue unrelated to autoDispose.

Dependencies are cleared after the provider rebuilds. But if a provider isn't listened, the provider won't rebuild and therefore will not remove its dependencies.

This will require changing how dependencies are cleared, which is a big change. 
That second issue will be fixed as part of https://github.com/rrousselGit/riverpod/issues/1253
